server/pbenchinacan/run-pbench-in-a-can

```diff
@@ -1,24 +1,125 @@
+export PB_COMMON_IMAGE_PULL_POLICY=${PB_COMMON_IMAGE_PULL_POLICY:-"always"}
 export PB_ELASTICSEARCH_IMAGE_PULL_POLICY="${PB_ELASTICSEARCH_IMAGE_PULL_POLICY:-${PB_COMMON_IMAGE_PULL_POLICY}}"
 export PB_POSTGRESQL_IMAGE_PULL_POLICY="${PB_POSTGRESQL_IMAGE_PULL_POLICY:-${PB_COMMON_IMAGE_PULL_POLICY}}"
 export PB_KEYCLOAK_IMAGE_PULL_POLICY="${PB_KEYCLOAK_IMAGE_PULL_POLICY:-${PB_COMMON_IMAGE_PULL_POLICY}}"
 export PB_SERVER_IMAGE_PULL_POLICY="${PB_SERVER_IMAGE_PULL_POLICY:-${PB_COMMON_IMAGE_PULL_POLICY}}"
 
 export PB_DASHBOARD_DIR="${PB_DASHBOARD_DIR:-${PWD}/dashboard/build/}"
 
-envsubst < $(dirname ${0})/pod.yml | podman play kube -
+podman pod create --name ${PB_POD_NAME} --hostname pbenchinacan -p 9200:9200 -p 5432:5432 -p 8090:8090 -p 8080:8080
+
+podman run \
+  --rm \
+  --name "${PB_POD_NAME}-elasticsearch" \
+  --pod ${PB_POD_NAME} \
+  --workdir /usr/share/elasticsearch \
+  --pull ${PB_ELASTICSEARCH_IMAGE_PULL_POLICY} \
+  --umask 0000 \
+  --restart no \
+  --pids-limit 0 \
+  -e PATH=/usr/share/elasticsearch/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin \
+  -e HOME=/root \
+  -e TERM=xterm \
+  -e container=oci \
+  -e discovery.type=single-node \
+  -e ES_JAVA_OPTS="-Xms1g -Xmx1g"\
+  -e xpack.security.enabled=false \
+  -e ES_PORT=9200 \
+  -d \
+  ${PB_CONTAINER_IMAGE_PREFIX}/${PB_ELASTICSEARCH_IMAGE_NAME}:${PB_ELASTICSEARCH_IMAGE_TAG} \
+  /bin/tini -- /usr/local/bin/docker-entrypoint.sh eswrapper
+
+podman run \
+  --rm \
+  --pod ${PB_POD_NAME} \
+  --name "${PB_POD_NAME}-postgres" \
+  --workdir /opt/app-root/src \
+  --pull ${PB_POSTGRESQL_IMAGE_PULL_POLICY} \
+  --umask 0000 \
+  --user 26:26 \
+  --restart no \
+  --pids-limit 0 \
+  --entrypoint container-entrypoint \
+  -e POSTGRESQL_USER=pbenchcontainer \
+  -e POSTGRESQL_PASSWORD=pbench \
+  -e POSTGRESQL_DATABASE=pbenchcontainer \
+  -e HOME=/var/lib/pgsql \
+  -e CONTAINER_SCRIPTS_PATH=/usr/share/container-scripts/postgresql \
+  -e APP_ROOT=/opt/app-root \
+  -e POSTGRESQL_VERSION="13" \
+  -e PGUSER=postgres \
+  -e STI_SCRIPTS_URL=image:///usr/libexec/s2i \
+  -d \
+  ${PB_CONTAINER_IMAGE_PREFIX}/${PB_POSTGRESQL_IMAGE_NAME}:${PB_POSTGRESQL_IMAGE_TAG} \
+  run-postgresql -p 5432
+
+podman run \
+  --rm \
+  --pod ${PB_POD_NAME} \
+  --name "${PB_POD_NAME}-keycloak" \
+  --pull ${PB_KEYCLOAK_IMAGE_PULL_POLICY} \
+  --umask 0000 \
+  --restart no \
+  --pids-limit 0 \
+  -e KEYCLOAK_ADMIN=admin \
+  -e KEYCLOAK_ADMIN_PASSWORD=admin \
+  -d \
+  ${PB_CONTAINER_IMAGE_PREFIX}/${PB_KEYCLOAK_IMAGE_NAME}:${PB_KEYCLOAK_IMAGE_TAG} \
+  start-dev --health-enabled=true --http-port=8090
+
+source server/pbenchinacan/load_keycloak.sh
+
+# We expect the host dashboard directory already exists, however,
+# in case it is missing we will create an new empty directory.
+# Note: if the dashboard directory is not present Pbench Server runs
+# without a dashboard deployment.
+mkdir -m 0755 -p ${PB_DASHBOARD_DIR}
+
+podman run \
+  --rm \
+  --pod ${PB_POD_NAME} \
```

Comment:
We should run outside the pod with host networking so that we can talk to the pod's services it offers.

This is currently how we are planning to deploy in production, and are already using in staging.

Reply:
If we run outside the pod, won't that limit us to a single deployment per node?

Perhaps that's not a problem, especially if Mr. Jenkins never runs more than one job on a given executor at a time, but encapsulating the network seemed like a good idea.

Reply:
> We should run outside the pod with host networking so that we can talk to the pod's services it offers.

You mean pbench-server container should not be a part of the same pod that contains all our microservices? And that it should find the service over the host network?

Reply:
I think it's more about breaking up "the pod" and having containers working through host networking. In real deployment, we don't expect Elastic, Keycloak, or PostgreSQL to be "in a pod" ... they're completely independent containers (or VMs or even bare metal) somewhere else in the datacenter, and we don't care as long as we have host/port for each.

Reply:
I don't think it's likely to be a big deal, but we do want to mimic the expected production environment, and using host networking is probably best; in which case operating as "a pod" probably isn't helping us so much as confusing the relationship between test and reality.